File: shell/src-tauri/src/server/bridge.rs
//! Versioned file IPC with the REAPER Lua adapter.
//!
//! The encoding mirrors `urllib.parse.quote(value, safe="")`: RFC 3986's
//! unreserved bytes stay literal; every other UTF-8 byte is percent encoded.

use std::{fs, path::PathBuf};

use percent_encoding::percent_decode_str;

pub const PROTOCOL_VERSION: u8 = 1;

pub fn encode_fields(fields: &[&str]) -> String {
    fields.iter().map(|field| percent_encode(field)).collect::<Vec<_>>().join("|")
}

pub fn decode_fields(line: &str) -> Vec<String> {
    line.trim_end_matches(['\r', '\n']).split('|').map(|field| percent_decode_str(field).decode_utf8_lossy().into_owned()).collect()
}

/// Mirrors `urllib.parse.quote(value, safe="")` - also reused by
/// `guide_service.rs::preview_url`, which needs the identical encoding for
/// a filename embedded in an audio URL (`GuideService.cs`'s original used
/// the same `Uri.EscapeDataString`-equivalent call for both).
pub fn percent_encode(value: &str) -> String {
    let mut encoded = String::new();
    for byte in value.bytes() {
        if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'.' | b'_' | b'~') { encoded.push(byte as char); }
        else { encoded.push_str(&format!("%{byte:02X}")); }
    }
    encoded
}

pub struct BridgeClient { session_dir: PathBuf, counter: u64, event_offset: u64 }

impl BridgeClient {
    pub fn new(session_dir: impl Into<PathBuf>) -> Result<Self, String> {
        let session_dir = session_dir.into();
        fs::create_dir_all(session_dir.join("commands")).map_err(|error| format!("Could not create bridge command directory: {error}"))?;
        Ok(Self { session_dir, counter: 0, event_offset: 0 })
    }
    pub fn send(&mut self, action: &str, fields: &[&str]) -> Result<PathBuf, String> {
        let target = self.session_dir.join("commands").join(format!("{:08}.cmd", self.counter));
        self.counter += 1;
        let temporary = target.with_extension("tmp");
        let mut values = vec![PROTOCOL_VERSION.to_string(), action.to_string()]; values.extend(fields.iter().map(|value| (*value).to_string()));
        let refs = values.iter().map(String::as_str).collect::<Vec<_>>();
        fs::write(&temporary, format!("{}\n", encode_fields(&refs))).map_err(|error| format!("Could not write bridge command: {error}"))?;
        fs::rename(&temporary, &target).map_err(|error| format!("Could not activate bridge command: {error}"))?;
        Ok(target)
    }
    pub fn read_events(&mut self) -> Vec<String> {
        let path = self.session_dir.join("events.log");
        let Ok(bytes) = fs::read(&path) else { return vec![]; };
        let offset = usize::try_from(self.event_offset).unwrap_or(usize::MAX).min(bytes.len());
        self.event_offset = bytes.len() as u64;
        String::from_utf8_lossy(&bytes[offset..]).lines().filter(|line| !line.is_empty()).map(str::to_string).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::{decode_fields, encode_fields};
    #[test]
    fn matches_python_quote_safe_empty_for_unicode_and_protocol_delimiters() {
        assert_eq!(encode_fields(&["hello world", "A|B", "é", "~_.-"]), "hello%20world|A%7CB|%C3%A9|~_.-");
        assert_eq!(decode_fields("hello%20world|A%7CB|%C3%A9|~_.-\n"), vec!["hello world", "A|B", "é", "~_.-"]);
    }
}

File: shell/internal/bridge/bridge.go
// Package bridge implements the versioned, file-based IPC contract with the
// REAPER Lua launcher. Its percent encoding deliberately matches Python's
// urllib.parse.quote(value, safe="") byte-for-byte.
package bridge

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

const ProtocolVersion = 1

func EncodeFields(fields []string) string {
	encoded := make([]string, len(fields))
	for index, field := range fields {
		encoded[index] = PercentEncode(field)
	}
	return strings.Join(encoded, "|")
}

func DecodeFields(line string) ([]string, error) {
	parts := strings.Split(strings.TrimRight(line, "\r\n"), "|")
	decoded := make([]string, len(parts))
	for index, part := range parts {
		value, err := percentDecode(part)
		if err != nil {
			return nil, err
		}
		decoded[index] = value
	}
	return decoded, nil
}

func PercentEncode(value string) string {
	const hex = "0123456789ABCDEF"
	var builder strings.Builder
	for _, byteValue := range []byte(value) {
		if byteValue >= 'a' && byteValue <= 'z' || byteValue >= 'A' && byteValue <= 'Z' || byteValue >= '0' && byteValue <= '9' || byteValue == '-' || byteValue == '.' || byteValue == '_' || byteValue == '~' {
			builder.WriteByte(byteValue)
			continue
		}
		builder.WriteByte('%')
		builder.WriteByte(hex[byteValue>>4])
		builder.WriteByte(hex[byteValue&0x0f])
	}
	return builder.String()
}

type Client struct {
	mu          sync.Mutex
	sessionDir  string
	counter     uint64
	eventOffset int64
}

func New(sessionDir string) (*Client, error) {
	if err := os.MkdirAll(filepath.Join(sessionDir, "commands"), 0o755); err != nil {
		return nil, fmt.Errorf("could not create bridge command directory: %w", err)
	}
	return &Client{sessionDir: sessionDir}, nil
}

func (c *Client) Send(action string, fields []string) (string, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	target := filepath.Join(c.sessionDir, "commands", fmt.Sprintf("%08d.cmd", c.counter))
	c.counter++
	values := append([]string{fmt.Sprint(ProtocolVersion), action}, fields...)
	temporary := target + ".tmp"
	if err := os.WriteFile(temporary, []byte(EncodeFields(values)+"\n"), 0o600); err != nil {
		return "", fmt.Errorf("could not write bridge command: %w", err)
	}
	if err := os.Rename(temporary, target); err != nil {
		return "", fmt.Errorf("could not activate bridge command: %w", err)
	}
	return target, nil
}

func (c *Client) ReadEvents() ([]string, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	bytes, err := os.ReadFile(filepath.Join(c.sessionDir, "events.log"))
	if os.IsNotExist(err) {
		return []string{}, nil
	}
	if err != nil {
		return nil, err
	}
	offset := c.eventOffset
	if offset < 0 || offset > int64(len(bytes)) {
		offset = int64(len(bytes))
	}
	c.eventOffset = int64(len(bytes))
	text := string(bytes[offset:])
	result := []string{}
	for _, line := range strings.Split(text, "\n") {
		if line != "" {
			result = append(result, strings.TrimSuffix(line, "\r"))
		}
	}
	return result, nil
}

func percentDecode(value string) (string, error) {
	bytes := make([]byte, 0, len(value))
	for index := 0; index < len(value); index++ {
		if value[index] != '%' {
			bytes = append(bytes, value[index])
			continue
		}
		if index+2 >= len(value) {
			return "", fmt.Errorf("invalid percent escape")
		}
		hi, okHi := hexValue(value[index+1])
		lo, okLo := hexValue(value[index+2])
		if !okHi || !okLo {
			return "", fmt.Errorf("invalid percent escape")
		}
		bytes = append(bytes, hi<<4|lo)
		index += 2
	}
	return string(bytes), nil
}
func hexValue(value byte) (byte, bool) {
	if value >= '0' && value <= '9' {
		return value - '0', true
	}
	if value >= 'a' && value <= 'f' {
		return value - 'a' + 10, true
	}
	if value >= 'A' && value <= 'F' {
		return value - 'A' + 10, true
	}
	return 0, false
}
